### PR TITLE
Feature/serialisation

### PR DESCRIPTION
**Pull request by elgator**
Hi,
This is a proposed implementation of `to_dict()` game serialisation.
Top level `Game.to_dict()` chain-calls other classes.
Dict structure is described in the `Game.to_dict()` doc string.

### CODE REVIEW

**Comment by almaslov** [comment]
Hello!
I think serialization is more UI- or network-specific feature, so model classes should not be aware of it.
In my opinion, it's more convenient to hold serialization logic in a different module. It would keep the core components code clearer if we add other types of serialization, e.g. full state or state observed by a specific player. Also, it allows serialization logic to be more compact and not scattered throughout several modules.

**Comment by D-Mantrid** [comment]
Agree!
Let's create, say, `formatting.py` module that will contain a DictFormatter (or JSONFormatter) class doing all the job with Game object passed as a parameter.

**Comment by elgator** [comment]
Hi,
I object. If I move this serialisation logic into a separate module then it will have to know all the internal details of the model. All properties and connections. And if one decides to change the model, he will have to amend this new module as well.

I understand the argument of "other types of serialization". So I think the better solution would be to serialise everything, without filtering. And another module, that keeps communication with the front, would pass along only the needed part.

**Comment by almaslov** [comment]
On one hand, I agree, that it could break the incapsulation for some classes. On the other hand, our core classes interface is completely public. So when creating dict representation for an object, we add one more public method with no additional information. 
Moreover, the structure of the total dict perfectly repeats the structure of the game objects relationship, thus serializer should know it anyway to filter.
Also, (I understand it won't hit performance dramatically) there's good С++ mantra "you don’t pay for what you don’t use", which in this case I would reword as "you don't iterate over data you don't want to serialize".

**Comment by D-Mantrid** [comment]
There are some points to mention:
- The bigger matter is where we place functionality. A separate `formatter` module could have kept all the task for serializing/representing/printing objects inside itself. By doing this, we follow the Single Responsibility Principle and remove the responsibility from the core to know how to represent its state because it's not its responsibility.

- Now,
  > If I move this serialisation logic into a separate module then it will have to know all the internal details of the model. All properties and connections. And if one decides to change the model, he will have to amend this new module as well.

  It's ok for a supplementary component to be slightly dependent on the core component as long as the core component doesn't have to know about the former. We could provide the core with some interface level accessible for applicants but, like @almaslov said, in Python we often have all interfaces open (we have a lot of differences in Python from the other world, actually :slightly_smiling_face:). All in all, as our business logic is not stable yet, it's naive to think that its interface or other components will be more stable as they depend on business level stability by definition.

- Either with or without `.to_dict` interfaces, the formatter already knows about internal details of core objects it operates as it chooses what content to format and how to format. But at least we don't obligate next level components to know and operate these internals. To make it better, we can just let the formatter use only `Game.to_dict` interface. However, **why don't we want to use the formatter as this very interface?**
  At least while it belongs to the core domain...

- After all, we can fall into lie with unproper `to_dict` name (see Deck for instance which representation is not a dict).

My suggestion would be keeping one interface method left in a main core class (i.e. `Game.get_dict_state`) as it already knows about its vassals and make it usable in Formatter classes.

**Comment by elgator** [comment]
Ok, I will make a Formatter class.

File: src/formatter.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        output['infect_deck_discard'] = DictFormatter.deck_to_dict(game.infect_deck)
+        output['diseases'] = {colour: DictFormatter.disease_to_dict(disease) for colour, disease in
+                              game.diseases.items()}
+        output['cities'] = {name: DictFormatter.city_to_dict(city) for name, city in game.city_map.items()}
+        output['infection_rate'] = game.infection_rate
+        output['epidemic_count'] = game.epidemic_count
+
+        return output
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (elgator) [reply src/formatter.py:56]: Well, staticmethod can't be overridden. I did not expect that.

[the remark this answers is not in the document]
Comment (almaslov) [reply src/formatter.py]: Actually, it can. The one important thing is how you call it. If you expect static/class method can be overridden in a subclass, then it's better to call it via 'self' or 'cls' object.
In our previous case when using the 'staticmethod' decorator you loose information about the current class type, so you have to hardcode the current class name. In practice, that means, that if you want to, let's say, override card_to_dict to serialize just city name, you should also override all methods that invoke card_to_dict so they use overridden version.
When calling 'cls.card_to_dict', the 'cls' is not the class where the method defined, it's the type of object whose method is called.

File: src/city.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def __init__(self, name, colour):
         self.has_lab = False
         self.colour = colour
         self.infection_levels = {}
-        self.distance = 999
+        self.distance = 999 # TODO unused?
```
Comment (D-Mantrid) [inline src/city.py:22]: Yes, unused.
To be removed in the next cleanup branch.
Nothing to do here right now.

File: src/game.py [hunk 3]
```diff
@@ -34,7 +34,7 @@ def __init__(self):
         self.infection_rate = None
         self.infection_rates = []
         self.epidemic_count = 0
-        self.diseases = {}
+        self.diseases = {} # TODO why is it a dict? Disease has colour property
```
Comment (D-Mantrid) [inline src/game.py:37]: We use colour names as key labels to get an exact disease object e.g. when infecting cities (we use the colour of the drawn card).
Nothing to do here.

File: src/player.py [hunk 4]
```diff
@@ -11,12 +11,12 @@ def __str__(self):
...
 class Player:
     def __init__(self, name):
-        self.game = None
+        self.game = None # TODO is it used?
         self.location = None
         self.action_count = 0
         self.hand = []
         self.name = name
-        self.controller = None
+        self.controller = None # TODO is it used?
```
Comment (D-Mantrid) [inline src/player.py:19]: Obsolete. To be removed in the next cleanup PR.

File: src/player.py [hunk 5]
```diff
@@ -11,12 +11,12 @@ def __str__(self):
 
 class Player:
     def __init__(self, name):
-        self.game = None
+        self.game = None # TODO is it used?
```
Comment (D-Mantrid) [inline src/player.py:14]: Of course, e.g. in player actions and checks.
And after all each Characrter is a part of the Game.

File: src/formatter.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+class DictFormatter:
+    '''
+    Provides methods of serialisation of Game instance for various viewpoints
+    '''
+
+    def __init__(self, game):
+        self.game = game
+
+    def to_dict(self):
```
Comment (D-Mantrid) [inline src/formatter.py:11]: Do we really need this method along with `__init__` method now?

File: src/formatter.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        return {'name': card.name, 'colour': card.colour}
+
+    @classmethod
+    def deck_to_dict(cls, deck):
+        """
+        :param deck: Deck object
+        Does not return still uncovered cards
+        :return: dict (id => card dict), card objects are copied
+        """
+        return {id: DictFormatter.card_to_dict(card) for id, card in enumerate(deck.discard)}
```
Comment (D-Mantrid) [inline src/formatter.py:73]: Do we use need to use `DictFormatter` as `cls` here? If so, it's better to change it.

File: src/formatter.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        :action_count: int
+        :hand: dict of (id => Card dicts)
+            :name: str
+            :colour: str
+        """
+        output = {}
+        output['name'] = player.name
+        output['location'] = player.location
+        output['action_count'] = player.action_count
+        output['hand'] = {id: DictFormatter.card_to_dict(card) for id, card in enumerate(player.hand)}
```
Comment (D-Mantrid) [inline src/formatter.py:124]: Same question with `DictFormatter` vs `cls`.

File: src/formatter.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        :name: str
+        :location: str
+        :action_count: int
+        :hand: dict of (id => Card dicts)
+            :name: str
+            :colour: str
+        """
+        output = {}
+        output['name'] = player.name
+        output['location'] = player.location
```
Comment (D-Mantrid) [inline src/formatter.py:122]: I see a potential bug here which is also uncovered by tests.

File: src/formatter.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
+
+
+class DictFormatter:
```
Comment (D-Mantrid) [inline src/formatter.py:3]: Do we need exactly a `dict` converter?
I mean, some game components can be conveniently represented as lists or other primitives...
